src/bindings/python/flux/cli/plugin.py

```diff
@@ -0,0 +1,180 @@
+##############################################################
+# Copyright 2024 Lawrence Livermore National Security, LLC
+# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+#
+# This file is part of the Flux resource manager framework.
+# For details, see https://github.com/flux-framework.
+#
+# SPDX-License-Identifier: LGPL-3.0
+##############################################################
+
+import glob
+from abc import ABC
+from os import getenv
+
+from flux.conf_builtin import conf_builtin_get
+from flux.importer import import_path
+
+
+class CLIPluginOption:
+    """Wrap Argparse's add_argument method with a class
+
+    Args:
+
+        name (str): Long option string being added (must begin with ``--``).
+        prefix (str): By default, ``--ex-`` is added as a prefix to ensure
+            arguments are namespaced. However, other flux-framework projects
+            can override this (as can other projects, though it is not
+            advised).
+
+    Other keyword arguments, except ``dest=`` are passed along to
+    ``ArgumentParser.add_argument`` unchanged. ``dest=`` is a special case,
+    where dashes are replaced with underscores.
+    """
+
+    def __init__(self, name, prefix="--ex-", **kwargs):
+        if not name.startswith("--"):
+            raise ValueError("Plugins must register only long options.")
+        if prefix:
+            if "dest" not in kwargs:
+                kwargs["dest"] = name[2:].replace("-", "_")
+            self.name = name.replace("--", prefix)
+        self.__dict__.update(kwargs)
+
+    def get(self):
+        """Return a dict of keyword arguments for argparse"""
+        return self.__dict__
+
+
+class CLIPlugin(ABC):  # pragma no cover
+    """Base class for a CLI submission plugin
+
+    A plugin should derive from this class and implement one or more
+    base methods (described below)
+
+    Attributes:
+        prog (str): command-line subcommand for which the plugin is active,
+            e.g. "submit", "run", "alloc", "batch", "bulksubmit"
+    """
+
+    def __init__(self, prog):
+        self.prog = prog
+        if prog.startswith("flux "):
+            self.prog = prog[5:]
+        self.options = []
+
+    def add_option(self, option):
+        """Allow plugin to add additional options to itself
+
+        Plugins use:
+        >>> self.add_option(CLIPluginOption("--longopt", action="store_true", help="Help."))
```

Comment:
I would slightly prefer if the user did not have to instantiate the `CLIPluginOption()`. You may have a good reason for this, but why not:

```Python
    def add_option(name, prefix="--ex-", **kwargs):
        self.options.append(CLIPluginOption(name, prefix, **kwargs))
```

This is  a bit more user friendly interface IMO and doesn't require the user to import or use the `CLIPluginOption` class in the common case. However, you may have tried this and run into some issue I don't see.
```